### ISSUE TITLE
Per notebook MANIFESTS

### ISSUE THREAD

**Issue by Keno** [JuliaLang/IJulia.jl#673]
Now, that 0.7 is getting closer, it may make sense to start thinking about how notebooks interact with the new package manager. I had discussed with @StefanKarpinski and @KristofferC that it would be great if notebooks could embed a MANIFEST and thus if you send somebody a notebook they could automatically load everything with the correct versions. Doing something like this would require figuring out where to store the information, how to hook it up to Pkg3 and probably require some UI work as well.

**Comment by stevengj** [comment]
maybe with the [contents API](http://jupyter-notebook.readthedocs.io/en/stable/extending/contents.html)

**Comment by StefanKarpinski** [comment]
I believe that what's needed is a "environment protocol": i.e. instead of needing to actually have a project file and/or manifest file present, or a package directory, or load path array, one just needs to implement the environment protocol. Then the IJulia package can implement the protocol for notebooks that have environment information stored in them and voila, each notebook has its own environment. However, I think that work is a 1.x kind of thing: we now generally understand what the protocol needs to look like; the next step is to factor out the protocol part in such a way that the three kinds of environments that we already support are implementations of this protocol; after that we allow a notebook to implement the environment protocol as well. The main thing to consider at this point is how to allow for extension in the future. Where is the hook? Do we have a `Base.PACKAGE_ENVIRONMENT` variable, which, if set, overrides the `LOAD_PATH` lookup? Or do we have some special name which can be put into the `LOAD_PATH` that causes loading to talk to the notebook instead?

**Comment by jlperla** [comment]
A long-run solution is great to automatically embed the manifest/etc. But is it possible to have a short-term patch requiring a manual call to load something in the notebook itself?  That is, something along the lines of
```
Pkg.setmanifest("Manifest.toml") #i.e., local to the notebook
using MyLib #i.e., the kernel is using the `Manifest.toml` now
```
Or maybe this is already possible with some of the `Pkg3` commands in Jupyter?

**Comment by StefanKarpinski** [comment]
🤷‍♂️  maybe?

**Comment by KristofferC** [comment]
Can't you just `activate` the dur of the notebook? Then that notebook will use a separate environment that will be stored next to he notebook.

**Comment by jlperla** [comment]
To make sure I understand this, you think I may just be able to put a `Manifest.toml` in the notebook directory, then I should just need to run:
```julia
Pkg.activate(".")
using MyLib
```

If that is correct, I can try to have someone test it when IJulia is sufficiently stable with 0.7

**Comment by KristofferC** [comment]
You need a Project file as well. But yes if you do

```
Pkg.activate(".")
```

and then go wild with adding packages, those will be recorded in `Project.toml` and `Manifest.toml` along the notebook, and if you send these files to someone else, they can do

```
Pkg.activate(".")
Pkg.instantiate()
```

to install all the packages at the version you used them.

**Comment by tkf** [comment]
If opening a notebook can instantiate arbitrary `Manifest.toml` (which may contain arbitrary `repo-url`), isn't it a security hole?  Isn't it also incompatible with the security model of Jupyter notebook (= trust if you execute it)?

How about adding a simple function that uploads `Project.toml` and `Manifest.toml` to gist and then call `IJulia.load_string` to inject something like

```julia
Pkg.activate("https://gist.github.com/.../...")
```

to a notebook cell?  Of course, `Pkg.activate` then has to support downloading `*.toml` when a URL is specified.  `Pkg.activate` can also check if those packages are from the known registries and prompt user if not.

Alternatively, I guess you can use [cell attachments](https://nbformat.readthedocs.io/en/latest/format_description.html#cell-attachments) to bundle `*.toml` into the notebook file but it would require the kernel and the server to be in the same machine.  For example, it won't work if you launch a Julia kernel on a HPC cluster via a Jupyter notebook server running on your laptop.

**Comment by jlperla** [comment]
I don't know jupyter all that well, but isn't the security controlled by how it is contained? You can load local files, run shell stuff, etc if it lets you?

Certainly being able to instantiate a local manifest is not the long-run solution, and will not work for all scenarios, but I don't think it is a security hole.

**Comment by vchuravy** [comment]
I now have this in my notebooks

```
using Pkg
Pkg.activate(@__DIR__)
Pkg.instantiate()
pkg"precompile"
```

**Comment by KristofferC** [comment]
Storing the Manifest + Project inside the notebook and have a button that does that would come a long way. There shouldn't be any security problems with that, it is just a convenience layer?

**Comment by tkf** [comment]
If you are willing to write a front-end extension I think that's great! I have no intention of stopping it.

**Comment by simonbyrne** [comment]
The notebook does include a certain amount of notebook-wide metadata, detailing the language and kernel. e.g.
```
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.0.0",
   "language": "julia",
   "name": "julia-1.0"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.0.0"
  }
```

It may be possible to insert and read the manifest information from there.

As far as a security model goes, one solution could be a confirmation dialog before installing any new package versions via `activate`.

**Comment by simonbyrne** [comment]
What if we added a function to IJulia which did something along the lines of what @vchuravy [suggested](https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-414078714), which could be included at the top of IJulia files?
```
using Pkg
function useproject(path=".")
    Pkg.activate(joinpath(@__DIR__,path))
    Pkg.instantiate()
    pkg"precompile"
end
```

**Comment by tkf** [comment]
It does not work when IJulia kernel and Jupyter server run in different machines.
https://bitbucket.org/tdaff/remote_ikernel/src/default/
https://github.com/ipython/ipython/wiki/Cookbook:-Connecting-to-a-remote-kernel-via-ssh

**Comment by StefanKarpinski** [comment]
At JupyterCon I spoke with a few Jupyter folks and their take was that trying to put this kind of metadata into notebooks was not the right direction to go—they've tried this with images and other things in the past and have come to feel that the "unit of distribution" should be a git repo, not a single notebook file. So it seems like the way to go here might be to have IJulia automatically activate the project in the git repo that it's in. After all, you are running the code in the notebook, so presumably you trust it. (As compared to just starting a Julia process in a directory, which may or may not mean that you trust the content of the directory enough to execute it.)

**Comment by stevengj** [comment]
IJulia doesn't know what notebook file (if any) it is executing — that information is not provided to the kernel.

**Comment by Keno** [comment]
> At JupyterCon I spoke with a few Jupyter folks and their take was that trying to put this kind of metadata into notebooks was not the right direction to go—they've tried this with images and other things in the past and have come to feel that the "unit of distribution" should be a git repo, not a single notebook file. So it seems like the way to go here might be to have IJulia automatically activate the project in the git repo that it's in. After all, you are running the code in the notebook, so presumably you trust it. (As compared to just starting a Julia process in a directory, which may or may not mean that you trust the content of the directory enough to execute it.)

If we go this way, I'd still like a way to package everything into a single file that you can email to somebody or share on JuliaBox (also have separate environments for every notebook on juliabox). If we I just want to share some code with somebody, I don't think we can expect the workflow to be "Go clone this git repo".

**Comment by jlperla** [comment]
> I don't think we can expect the workflow to be "Go clone this git repo".

I agree. Jupyter notebooks need to be able to be used self-contained in some sense.  Even the Jupiter interface is often around the "Upload" notebook interface.  

What about the ability to activate from a URL?  You could give it the project file and/or manifest, and it would enable copying jupyter around.  And if someone wanted to run the notebook in whatever global project they had in their current jupyter, they wouldn't need to use those cells?

**Comment by StefanKarpinski** [comment]
I'm just reporting what the Jupyter people (@Carreau if I recall correctly) told me which is that they are moving away from trying to make notebooks self-contained because it has not worked out as hoped. The simplest solution would seem to be serving a zip or tar file continaing a set of notebooks, resources used by the notebook and in our case, project and manifest files.

**Comment by Carreau** [comment]
Yes, we tend to try to think of (1 unit == 1 repository).The notebook as unit, espescially since you can now connect many notebook to same kernel make not much sens.

We haven't really figured out how to make all of the completely work, but generally trying to shove more into a notebook does not work. 

As said before a repository does not always work, but I don't think we can get a "one size fits all". There is always this tension between being able to manipulate things on the filesystem, and having everything being opaque and managed by Jupyter.

You could of course have an extension for jupyter that show "bundles" as an actual tree of files, but then you can't cd into it.

Maybe something along a fuse driver that expose a single file at some path, and repo structure in another ?

**Comment by Carreau** [comment]
@fperez would be interested in this discussion BTW, and I think we had pictures of a whitebord with all the different axes of what people want from notebook files.

**Comment by simonbyrne** [comment]
My experience is that embedding data in notebooks is a lost cause. e.g. the attachments feature is basically useless, since:
- you can't access the attachments from the kernel (https://github.com/JuliaLang/IJulia.jl/issues/625), so it's no use for embedding data.
- attachments are lost by nbconvert (https://github.com/jupyter/nbconvert/issues/699), so you can't even use it for embedding images if you want to, say, convert the notebook to a presentation.

**Comment by vchuravy** [comment]
I often enough ship notebooks to students who work in isolated (supercomputing) environments. So needing internet access to replicate a notebook would be an annoyance.

If we say that the unit is the git repository that is fine, but I really would like to enable a workflow where somebody can just grab a notebook itself.

To me Project/Manifest are very unlike pictures and other attachments. A notebook doesn't break just because a picture is missing, but it won't work without the correct manifest. So embedding that in the Metadata would be preferable.

I recently gave a workshop and it boiled down to having an environment per notebook in a different subfolder.

**Comment by StefanKarpinski** [comment]
Yes, I think you're right @tkoolen, but most of the time that directory is the root of a git repo and git repos do end up being the unit of reproducibility (although git trees actually work too).

**Comment by simonbyrne** [comment]
@tkf what does "it" refer to here?

> It does not work when IJulia kernel and Jupyter server run in different machines.

**Comment by Carreau** [comment]
> Yes, I think you're right @tkoolen, but most of the time that directory is the root of a git repo and git repos do end up being the unit of reproducibility (although git trees actually work too).

Yes, and usually it provides "collaboration" capability like synchronisation or something else. It's one way of abusing language,  saying that the unit of reproducibility/sharing is bigger than a notebook (and may not contain a notebook).

**Comment by tkf** [comment]
@simonbyrne `IJulia.useproject()` you suggested assumes that the kernel is running on the machine where the notebook is. This is not true in general since you can connect to remote kernel via ssh (say).  But maybe you can argue that is too exotic usecase so it is not worth supporting.

BTW, my suggestion https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-414033742 does not have these shortcomings you pointed out for the attachment approach:

> * you can't access the attachments from the kernel (#625), so it's no use for embedding data.
> * attachments are lost by nbconvert (jupyter/nbconvert#699), so you can't even use it for embedding images if you want to, say, convert the notebook to a presentation.

**Comment by simonbyrne** [comment]
Ah, thanks. That makes sense.

To be wholly self contained (i.e. avoid things like downloading manifests from gists), I think the only real solution is to do what @StefanKarpinski suggested and have an "environment API", so that you could specify something equivalent to the Manifest.toml in the first cell (perhaps in a less verbose format), along with a "snapshot" function to give the required output.

**Comment by tkf** [comment]
Manifest.toml has to hold rather big metadata (dependencies of dependencies) so my naive guess is that it's hard to squash it into a small notebook code cell.

> self contained (i.e. avoid things like downloading manifests from gists)

Why do you want to avoid network access while you need it to install packages anyway?  Also, you can include gist sha to the url.  Since git is immutable, it then becomes fully self-contained in the sense that the dependency tree is fully determined by a single notebook file.

**Comment by KristofferC** [comment]
The manifest is not only needed for installing packages but also to determine what you can load. Without it, you are blind. So putting that in a url seems like a bad idea.

**Comment by tkoolen** [comment]
Alright, so do people think it's actually desirable to have the `.toml` files embedded in the notebook at this point? I'd actually argue that: no it's not even desirable even if it is technically possible, because it'd be a very unconventional/magic way of working with Pkg.

Some thoughts:

* If the convention were to be to just put your notebooks in separate directories juxtaposed with their corresponding `.toml` files, and to explicitly add `import Pkg; Pkg.activate(@__DIR__); Pkg.instantiate()` at the start of the notebook, it would at least be very clear what's going on. 
* Should `File > Download as > Julia (.jl)` result in a `.jl` file that when run, always produces the same results as the notebook, no matter the directory to which it is downloaded? If so, how can that be reconciled with the previous point?
* One way to start Jupyter is to run `using IJulia; notebook(dir="/some/path")` from Julia. Perhaps there could be an extra `project` kwarg for that to mirror the `--project` Julia argument, with default value either `dir`, or `pwd()`, or `Base.active_project()`, passed along to the kernels. But another way is to run`jupyter notebook` from the command line, and `jupyter` is of course agnostic as to the kernel, so how do you make those two consistent?
* (I personally have `IJulia` in my default environment, so maybe it needn't be required to list it explicitly in the `Manifest.toml` for the notebook directory (thinking about NBInclude.jl here), but I could definitely be persuaded otherwise.)

**Comment by tkf** [comment]
> The manifest is not only needed for installing packages but also to determine what you can load. Without it, you are blind.

That's why I suggested:

> Pkg.activate can also check if those packages are from the known registries and prompt user if not.

Also, notebook already has using/import statements in it. It's very transparent what are going to be loaded (given that it uses registries you trust).

**Comment by tkf** [comment]
> But another way is to run jupyter notebook from the command line, and jupyter is of course agnostic as to the kernel, so how do you make those two consistent?

@tkoolen I think the default cwd of a jupyter kernel launched by jupyter notebook/lab is the directory of the notebook file.

**Comment by StefanKarpinski** [comment]
What's wrong with sending people a zip file of a directory?

**Comment by tkf** [comment]
@StefanKarpinski It does not work with some exotic usecase: https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-424907156. But I guess not supporting it makes sense since then simple `Pkg.activate("."); Pkg.instantiate()` just works.

**Comment by jlperla** [comment]
Sorry if this is an asinine and already rejected suggestion, but what about skipping Jupyter based modifications and rely entirely on the package manager?

That is, users have the option to
1. Follow https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-424933939 , etc.
   - Distribute the Manifest and Project with the notebook, and go `] activate .;` and/or `] activate .; instantiate` at the top of the notebook
   - Not self-contained notebook.  Possible problems with  https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-424907156 
2. Create a lightweight package without any code, and probably just a Manifest and Project files.  The notebook using the package is then completely self-contained.
   - Then at the top of the notebook they could say `] add MyNotebookProject; activate MyNotebookProject; instantiate` or whatever.  They would **not** do `using MyNotebookProject` because there is no code or reexporting.  Or if not registried, then `] add https://github.com/myproject/MyNotebookProject.jl`.
   - This would work with private repos, urls, local paths, private registries, etc....and the caching of the package manager would mean that it wouldn't necessarily require an internet condition and wouldn't require downloading if already cached - even if unregistered and added by URL.
   - Since the package would be installed wherever the kernel is running, it gives a lot of flexibility
   - The notebooks are self-contained, and in all likelihood the lightweight packages would be shared by a large number of notebooks on a site (since a super-set of the dependencies can be put into the file).

Is this an abuse of the package manager?  Can the registries (and uncurated registries) handle a proliferation of lots of small convenience packages, or would it break things?  Of course, this wouldn't really be one-project per notebook.  I realize this is inconvenient with the current `METADATA` based package registration, but I imagine that infrastructure could change.

**Comment by tkoolen** [comment]
@StefanKarpinski, re:

> What's wrong with sending people a zip file of a directory?

I think that's the way to go.

@tkf, re: 

> I think the default cwd of a jupyter kernel launched by jupyter notebook/lab is the directory of the notebook file.

That's true. But regardless of what `pwd()` is, an important question is still what `Base.active_project()` should be if you:

1. run `jupyter notebook` from `/some/path` and open a Julia notebook (possibly in a subdirectory).
2. run `using IJulia; notebook(dir="/some/path")` from Julia started in an arbitrary directory and with an arbitrary `Base.active_project()`, and open the same notebook.

I think that calling `Base.active_project()` from the first cell of the notebook should at least return the same directory in both cases, but if we really believe in 'unit == directory', **maybe it should be equal to `joinpath(@__DIR__, "Project.toml")` for the running notebook** instead of the current `~/.julia/environments/v1.0/Project.toml`. Either that or every notebook needs an `IJulia.useproject()` as the first cell as in https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-424199315, which has the advantage that it's clear what's going on, but would still be unfortunate boilerplate to have in (almost) every notebook.

Using `joinpath(@__DIR__, "Project.toml")` as the default `Base.active_project()` for a notebook also addresses https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-424216750 I think. But if this proposed default value is not what you want, you can always use `Pkg.activate` in a notebook cell to change it as desired, for exotic use cases.

**Comment by tkf** [comment]
> Create a lightweight package without any code, and probably just a Manifest and Project files. The notebook using the package is then completely self-contained.

@jlperla Yeah that's essentially equivalent to my suggestion https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-414033742.  Though I don't think you need to turn it into a package.  With the current infrastructure, you can already do:

```julia
run(`git clone $URL workspace`)
cd("workspace")  # has Project.toml [and Manifest.toml] in it
using Pkg
Pkg.activate(".")
Pkg.instantiate()
```

**Comment by tkf** [comment]
@tkoolen I don't think automatically activating an environment works unless it is automatically instantiated.  However, automatic instantiation deviates from Jupyter's security model (= you trust the notebook if you ever run it) https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-414096386.  You could put `using Pkg; Pkg.instantiate()` in the first cell but then single `IJulia.useproject()` would do the same and more explicit.

> Using `joinpath(@__DIR__, "Project.toml")` as the default Base.active_project() for a notebook also addresses https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-424216750

No, I don't think so, because in the scenario I described, the notebook file and the kernel are running in different machines (e.g., `jupyter lab` in your laptop and IJulia kernel on some cloud compute node).  `pwd()` of the remote IJulia kernel won't reflect the location of local notebook path (whose directory may not exist in the mote machine).

**Comment by tkf** [comment]
Other than the "exotic" `remote_ikernel` usage, I wonder how the approach with `*.ipynb` and `*.toml` in a directory plays with the realtime collaboration in Jupyter, something like (now deprecated) [jupyterlab-google-drive](https://github.com/jupyterlab/jupyterlab-google-drive).  It looks like the notebooks do not exist as a local JSON file anymore in this case and you can't have `*.toml` files besides them https://github.com/jupyterlab/jupyterlab-google-drive/issues/39.

**Comment by StefanKarpinski** [comment]
> It does not work with some exotic usecase: #673 (comment).

I don't mean loading from a zip file, I mean just sending someone a zip file and then they unzip it. The only real requirement here seems to be being able to send people a single file. I don't see why having a single file on the local filesystem where it's running is required.

**Comment by tkf** [comment]
> just sending someone a zip file and then they unzip it

@StefanKarpinski I'm just saying that this is too simplistic approach to cover other usages in Jupyter notebook/lab.  The kernel may be running in a different machine (e.g., `remote_ikernel` https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-424907156) and file system may be virtualized (e.g., jupyterlab-google-drive https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-424979926).

**Comment by StefanKarpinski** [comment]
It does not seem like Jupyter currently has the features to support this kind of thing. I don't think it should really fall on us to try to work around such limitations. The appropriate path forward seems like it would be conveying what we would need to do what we want to do.

**Comment by tkf** [comment]
> It does not seem like Jupyter currently has the features to support this kind of thing.

Jupyter has [`set_next_input`](https://jupyter-client.readthedocs.io/en/stable/messaging.html) protocol (invokable via `IJulia.load_string`) to support implementing what I suggested in https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-414033742

**Comment by vchuravy** [comment]
In discussion in slack I cam up with the following:

```
import LibGit2
using Pkg
env = joinpath(@__DIR__, "DemoEnvironment")
isdir(env) || LibGit2.clone("https://gist.github.com/2e4ebf0df689f4409d4341d366c89f15.git", env)
repo = LibGit2.GitRepo(env)
LibGit2.checkout!(repo, "708b17c88a89a88f08f4f1070e04b2a32974b1b7", force = true)
Pkg.activate(env)
pkg"instantiate"
pkg"precompile"
```

While quite verbose it encapsulates what I want from Manifest integration in Jupyter.
1. The ability to just send a notebook around/have someone download a notebook
2. Versioned dependencies that point to a "immutable" location

Kristoffer had the idea that we could maybe have something like `Pkg.activate("https://gist...", "sha")`that basically activates an anonymous environment and uses the merkle hash of gist repo + sha to cache and identify the environment.
I have wanted a *environment publish* command before that takes my current environment and uploads it so that I can share it with others for debugging reasons.

**Comment by StefanKarpinski** [comment]
I'm glad this lead to something that seems generally useful and also not Jupyter-specific.

**Comment by tkf** [comment]
Let me shamelessly point out that activating a remote repository is what I suggested in the very first post https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-414033742

**Comment by arnavs** [comment]
FYI, we've tagged a release of `QuantEcon/InstantiateFromURL.jl`, which implements the idea from @vchuravy above.

**Comment by jlperla** [comment]
To be clear, this first implementation is for a light repo with package and manifest. Which provides a solution for tightly controlled lecture notes /etc. The gist approach, which would be better for less formal setups, could be added as well if anyone is interested

**Comment by tkoolen** [comment]
I think https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-425306944 is missing a `Pkg.build()` step in order for things to be guaranteed to work starting from a clean slate. Would be nice not to have to do that every time you run the notebook though.

**Comment by KristofferC** [comment]
Instantiate builds the packages that got downloaded so don't think that is required.

**Comment by tkoolen** [comment]
I seem to recall cases when that didn't happen, but maybe that was just because the build had failed during an earlier `instantiate` call.

**Comment by arnavs** [comment]
@tkoolen FYI, the way we avoid rebuilding every time is to either (a) precompile the resources, for git refs that point to moving targets like `master`, or (b) version the resources using git tags, so something like `activate_github("arnavs/InstantiationTest", tag = "v0.1.0")` will never be updated.

**Comment by c42f** [comment]
Perhaps there's a very simple solution to this problem: treat the desired embedded environment metadata as code in the first executable cell. The question then becomes how to make it unobtrusive in the standard jupyter UI. It appears the UI doesn't do line wrapping, so there might also be a simple answer to that as well: base64 encode the toml files into a single line each.

The nice thing about this is that it's a solution for scripts which need to "come with their environment" just as much as jupyter notebooks. Then we'd just need a package `ProjectEnvironments` (or something) with a very simple and forward/backward compatible API which people could add manually, and which acts as the springboard into the well defined environment for the notebook.

Would this work or have I missed something?

**Comment by c42f** [comment]
I tried implementing this; there's a few goatchas but it looks like it will work. Gotchas include:
* We need to load [`CodeEnvironments`](https://github.com/c42f/CodeEnvironments.jl) (my working name for the package) from some default environment. Its API would need to be very forward and backward compatible.
* Mutable environments are somewhat problematic; you want the jupyter user to be able to add easily to the environment, but this conflicts with a desire to make them immutable and content addressed for the purposes of activating them from jupyter code. I think this can be managed in practice with some clear warnings but it's bit of a nasty wrinkle.

Generally there seems to be some impedance mismatch with `Pkg`, which is probably not a surprise given that I don't know a lot about `Pkg` ;-)  It does, however, offer a way to have per-notebook embedded manifests and project files.

**Comment by Keno** [comment]
Another datapoint: I'd like to be able to send people links to colab notebooks with in-built environments, but the unit they use is a file :)

**Comment by c42f** [comment]
I think my proposed solution/workaround would be ok for that. Would you be interested in it becoming a registered package? I'd need to think a bit more about the workflow and API, and probably involve Pkg people to know whether it's going to work out, or is fundamentally broken in some way. But I'm not sure whether to do that extra work yet.

**Comment by jlperla** [comment]
@Keno @c42f  If you have been using a solution like this, there is another use case to consider: getting notebook users able to *update* the project and manifest when necessary.  This has proven to be very important for our set of lecture notes, otherwise people effectively start copying around notebooks and using copies of them to edit for assignments.

After doing this for the last 8 months, my gut says that metadata in a notebook could become hellish to maintain and lead to all sorts of user issues wondering why they have the wrong versions of packages.  I used to be of the opinion that hidden metadata was the right way to go, but have reversed my stand completely.  On the other hand, I will *never* reverse my stand that notebooks have to execute self-contained from a single file and that copies around toml files is a terrible idea.

--------------------------------

For what it is worth, the approach we implemented from people's (i..e @vchuravy 's) suggestion in https://github.com/QuantEcon/InstantiateFromURL.jl/ has been very successful.  Basically,
1. Put a manifest and project file on github
2. At the top of the notebook you call the package to "activate" a particular version of it (e.g. a tagged release, a commit, or master).  We have been tagging versions, but you could do any of them including just commits
3. When that line is run, it looks in a local temporary `.project` file to see if the version of that package has been downloaded.  If not, it downloads, activates, and instantiates.  Otherwise it just activates.  The instantiation has been a very helpful step for ensuring people are using the right versions of the packages and it makes installation a joke.
4. If we ever need to update the set of packages, we can just tell people to bump a tag version (or hash tag) at the top of the notebook.

Take a look at https://github.com/QuantEcon/quantecon-notebooks-jl/blob/master/kalman.ipynb as an example, but basically all that is needed is the 
```julia
using InstantiateFromURL
activate_github("QuantEcon/QuantEconLecturePackages", tag = "v0.9.6");
```
at the top of the page.  The project and manifest are versioned in https://github.com/QuantEcon/QuantEconLecturePackages

Now, for those who don't need to have a mini repository, @vchuravy had the initial idea that this sort of package could have a simple utility to setup a `gist` instead.  https://github.com/QuantEcon/InstantiateFromURL.jl/issues/18

We didn't need it ourselves and couldn't put in the development time, but I think it is exactly the sort of thing that is needed for more lightweight package management.

... all of that is to say: before starting on any new solution, please see if the workflow in this package is solid and feel free to submit PRs for new features.  If enough people vet this solution, a variation on it might make sense in Pkg.jl or at least a more formally maintained package.

**Comment by c42f** [comment]
@jlperla That sounds like a great workflow for your use case. My reservation is that it's not self contained and requires supporting infrastructure which can't easily be updated by the end users. This is probably a good feature in your case where you're running a class with homogeneous package requirements.

On the other hand, I'm helping a group of somewhat nontechnical PhD students with heterogeneous data management and analysis tasks. My thought is that I should be able to give them jupyter notebooks (and normal scripts!) which have embedded self-contained environments. I'd also like them to be able to update package requirements as their needs change. But at the same time, have them well defined and embedded within the notebooks so that package requirements are somewhat resistant to user error (eg emailing a script and forgetting to add the Project and Manifest files).

**Comment by jlperla** [comment]
> (eg emailing a script and forgetting to add the Project and Manifest files).

Emailing project and manifest files around simply does not work.  I am completely with you.  And they make put the wrong ones with the wrong files.

> My thought is that I should be able to give them jupyter notebooks (and normal scripts!) which have embedded self-contained environments. I'd also like them to be able to update package requirements as their needs change.

I understand the goal of a "self-contained environment", but I would decouple that from a self-contained file.  Here are some usage scenarios:

> But at the same time, have them well defined and embedded within the notebooks so that package requirements are somewhat resistant to user error

- What  if you want to bump the versions of the notebooks that your students were using (which happens *all the time* with julia since the packages change frequently and are often broken)?   You can't have them just do manual package operations into the metadata because it is easy to get versions out of sync or to make mistakes
- What if they do package operations in the notebook and mess up versions?  How do they reset the notebook.
- What if they copy the notebook, start making their own changes... then you give them new ones with different and fixed packages but they get out of sync and don't realize they are working with an old set of hidden metadata.

These are the tip of the iceberg....  As I said, I used to think that this stuff belonged in the notebook but changed my tune completely after seeing usage scenarios.

> I'd also like them to be able to update package requirements as their needs change.

Having these things centrally managed is extremely helpful.  But I understand that having a full repo for the set of project/toml is a little heavy for most uses.

This is exactly why @vchuravy had originally suggested using a gist with some tools (which I will try to summarize below).  For us, having a consistent set of versions to bump was very nice but things don't need to have a full and controlled repository.

Basically, I think he had in mind https://github.com/QuantEcon/InstantiateFromURL.jl/issues/18  as a formalization of https://github.com/JuliaLang/IJulia.jl/issues/673#issuecomment-425306944

-  There would be a utility to for users to easily create the `gist` on their github account for a given
```
using InstantiateFromURL
hash = publish_gist(".") # by default, gets the local `project.toml and `manifest.toml` from the local file
# Could optionally pass in the github username, or use the github config to see it.
# e.g. hash = 2e4ebf0df689f4409d4341d366c89f15 
```
- Then, in a notebook you would have the hardcoded hash and put at the top:
```
using InstantiateFromURL
activate_gist("2e4ebf0df689f4409d4341d366c89f15") # optionally have a tag?
```
- If anyone wanted to update their gist, then could just call `publish_gist(".", hash)` to commit and push changes

.... or something along those lines.

**Comment by simonbyrne** [comment]
I've been using notebooks + toml in gists for a while, and while it works, there are some hassles

1) setting it up is a bit of a paint: you have to create the gist, then clone it back to the directory. Could be addressed by a script (though you would require a GitHub API key), but would be nicer if it could be done via Jupyter itself. Once set up though, pushing updating is easy.

2) all my gists end up being called "simonbyrne/Manifest.toml" (I assume because this is the file that appears first when sorted by ASCII code?). GitHub doesn't seem to provide a mechanism to rename them (you can change the comment that appears below, but not the name).

**Comment by arnavs** [comment]
Not sure if this helps, but the InstantiateFromURL package grabs repo tarballs (which don’t require an API key), and we store them (names salted with SHA hash) in a hidden directory from where the script is run. 

Could be different on the gist side, though.

**Comment by jlperla** [comment]
> Setting it up is a bit of a paint: you have to create the gist, then clone it back to the directory. Could be addressed by a script (though you would require a GitHub API key), but would be nicer if it could be done via Jupyter itself. Once set up though, pushing updating is easy.

I agree, and those sorts of scripts built into a package seem to be what Valentin was getting at.  I think it is a perfect case for a light package (which could ultimately become a feature of Pkg3 itself).  I am hesitant to say that we should have it in "jupyter" or IJulia since this is a more general problem than just jupyter notebooks.

If anyone wants to work on gist features, @arnavs  and I would be happy to merge them into the `InstantiateFromURL.jl` as a testbed

**Comment by c42f** [comment]
> These are the tip of the iceberg.... As I said, I used to think that this stuff belonged in the notebook but changed my tune completely after seeing usage scenarios.

These are all good points but come with strong assumptions that:
* You are teaching a large set of students, who all have to do more or less the same thing (ie, class assignments).
* You as the instructor are available to set up the version information in a centrally hosted location.

Consider instead that you are helping a group of nontechnical colleagues (students and lab staff) with their individual projects, each of which has different package requirements. This situation is a very different use case and I don't see how `InstantiateFromURL` can help with it.

**Comment by jlperla** [comment]
> This situation is a very different use case and I don't see how InstantiateFromURL can help with it.

Hence the suggestion to have gist based workflows from some people with simple for publishing. We didn't build it because of lack of time and not knowing requirements since we didn't need it. 

 My points are primarily about the difficulty of having relatively non-technical people manage project and manifests within the jupyter files themselves and all the things that can go wrong. 

The other thing to consider is that the students can use a base set of packages and then install additional ones with the by the build commands at the top of their own notebooks. 

But I could be wrong... Maybe there is some sort of technology that could make managing embedded package information within a notebook seamless and manageable. But it is hard to imagine without deep integration of both IJulia and Pkg3 (which there seems to be little appetite for).

**Comment by tkf** [comment]
@c42f FYI [`IJulia.load_string`](https://github.com/JuliaLang/IJulia.jl/blob/c02868e2be80ca5d7f7ffc9e81ee7740534b765b/src/IJulia.jl#L92-L99) seems to be a better option than `clipboard` when you are using it in Jupyter notebooks.

> * Mutable environments are somewhat problematic; you want the jupyter user to be able to add easily to the environment, but this conflicts with a desire to make them immutable and content addressed for the purposes of activating them from jupyter code.

I thought about how to address it. Here is an idea: put the following code with a hypothetical function `use_packages` in a hypothetical package `IJuliaPkg` at the top of the notebook:

```julia
using IJuliaPkg
use_packages(
    [
        "Plots",
        "DifferentialEquations",
    ],
)
```

which adds the packages in a plain environment, encode `Project.toml` and `Manifest.toml` in base64 or upload it to gist (hereafter I call the Julia object for it `$ENCODED_PROJECT`), and then replace the current cell with

```julia
using IJuliaPkg
use_packages(
    [
        "Plots",
        "DifferentialEquations",
    ],
    project = $ENCODED_PROJECT,
)
```

using `IJulia.load_string(..., true)`.  It should be easy to make `use_packages` idempotent; i.e., don't do anything other than `instantiate`+`activate` when the set of packages to be installed is identical to the one recorded in `Project.toml` in `$ENCODED_PROJECT`.  I think this lets you change the requirements of the notebook as you go.  That is to say, if you want to import `PyPlot`, go to the top of the notebook and edit it to

```julia
using IJuliaPkg
use_packages(
    [
        "Plots",
        "DifferentialEquations",
        "PyPlot",
    ],
    project = $ENCODED_PROJECT,
)
```

and then hit shift+enter which updates `$ENCODED_PROJECT`.

**Comment by c42f** [comment]
@tkf thanks, that's an excellent point. I had just assumed overwriting a code cell from the kernel was impossible!  With this in mind I think it's possible to have a self contained solution.

**Comment by stevengj** [comment]
Closed by #820.
